typescript-for-haskellers/article.md

```diff
@@ -0,0 +1,1002 @@
+// You should fully describe all types to not lose information.
+
+type A = {};
+type B = A & {x: number};
+
+let a: A = {};
+const b: B = {x: 1};
+a = b;
+a.x; // Property 'x' does not exist on type 'A'.
+```
+
+### Mapped types, Conditional types and Type families
+
+One useful type functionality of TypeScript are Mapped and Conditional types.
+They give us really flexible ability to modify existing types or create now one.
+
+#### Mapped types
+
+Mapped types allows you to create new types based on old types transforming them in some way.
+The syntax looks like `[set of field names]: type;`.
+For example, you can add optional modifier or readonly modifier to each field with `type Partial<T> = { [P in keyof T]?: T[P] }` and `type Readonly<T> = { readonly [P in keyof T]: T[P] }`.
+It is also possible to remove such modifiers using `-?` and `-readonly`.
+
+The `keyof` operator in the code below produces a union of properties name.
+Square brackets indicates computed names.
+And `in` operator goes over all field keys.
+
+`Partial` type adds `?` modifier for each property.
+
+```typescript
+type Partial<T> = { [P in keyof T]?: T[P] };
+
+type A = { x: number; y: number };
+type PartialA = Partia<A>;
+
+const a: A = { x: 1 }; // Property 'y' is missing in type '{ x: number; }' but required in type 'A'.
+const a: PartialA = { x: 1 }; // Ok
+```
+
+Let's try now to remove `?` and `readonly` modifier from all properties.
+
+```typescript
+type NoModifiers<T> = { -readonly [P in keyof T]-?: T[P] };
+
+type A = { readonly x?: number; readonly y?: number };
+type NoModifiersA = NoModifiers<A>;
+
+const a: A = { x: 1 }; // Ok
+a.x = 1; // Cannot assign to 'x' because it is a read-only property.
+
+const a: NoModifiersA = { x: 1 }; // Property 'y' is missing in type '{ x: number; }' but required in type 'NoModifiers<A>'
+const a: NoModifiersA = { x: 1, y: 2 }; // Ok
+a.x = 2; // Ok
+```
+
+Let's present some complex examples of mapped types usage.
+
+Defining type which contains string and numerical properties.
+
+```typescript
+type Point = {
+  x: number;
+  y: number;
+  0: string;
+}
+```
+
+Now, using mapped types we will create a type which will pick only some properties.
+Word `extends` here say that generic `K` will contain a subset of union properties keys `T`.
+And the resulting type will contain only properties from `K`.
+
+```typescript
+type Pick<T, K extends keyof T> = {
+  [P in K]: T[P];
+};
+
+type PointValues = Pick<Point, "x" | 0> // type PointValues = { x: number; 0: number }
+```
+
+The next example will be more complex.
+We will try to create the type which produce type with getters for all string properties.
+
+To implement it lets firstly define helper type `Getter`.
+Capitalize is a build in type which may be applied for the string type.
+So, for `T` which extends string we can produce getter specific name.
+
+```typescript
+type Getter<T extends string> = `get${Capitalize<T>}`;
```

Comment:
Do you need to capitalize? It doesn't seem relevant for illustrating any getter, and then you won't need to explain it.

Reply:
>Do you need to capitalize?

Without capitalize it will look like `getsomething`. And usually you need to somehow separate the words.\
I really like this example as it show how you can change types using `` and how complex mapped types can be

Reply:
Ok, sure. 👍